xrpl/utils/txn_parser/utils/order_book_parser.py

```diff
@@ -0,0 +1,191 @@
+    node_type = node["NodeType"]
+    if node_type == "CreatedNode":
+        return "created"
+    elif node_type == "ModifiedNode":
+        return "partially-filled"
+    else:  # node_type == "DeletedNode"
+        previous_fields = node.get("PreviousFields")
+        # a filled offer has previous fields
+        if previous_fields is not None:
+            return "filled"
+        # a cancelled offer has no previous fields
+        return "cancelled"
+
+
+def _derive_currency_amount(
+    currency_amount: Union[str, Dict[str, str]]
+) -> CurrencyAmount:
+    if isinstance(currency_amount, str):
+        return CurrencyAmount(currency="XRP", value=str(drops_to_xrp(currency_amount)))
+    else:
+        return CurrencyAmount(
+            currency=currency_amount["currency"],
+            issuer=currency_amount["issuer"],
+            value=currency_amount["value"],
+        )
+
+
+def _calculate_delta(
+    final_amount: CurrencyAmount,
+    previous_amount: CurrencyAmount,
+) -> str:
+    final_value = get_value(final_amount)
+    previous_value = get_value(previous_amount)
+    delta = final_value - previous_value
+    return str(delta)
+
+
+def _get_change_amount(
+    node: NormalizedNode,
+    side: Literal["TakerGets", "TakerPays"],
+) -> Optional[CurrencyAmount]:
+    new_fields = node.get("NewFields")
+    if new_fields is not None:
+        new_fields_amount = new_fields.get(side)
+        if new_fields_amount is not None:
+            return _derive_currency_amount(new_fields_amount)
+    final_fields = node.get("FinalFields")
+    previous_fields = node.get("PreviousFields")
+    if final_fields is not None:
+        final_fields_amount = final_fields.get(side)
+        if final_fields_amount is not None:
+            final_amount = _derive_currency_amount(final_fields_amount)
+            if previous_fields is not None:
+                previous_fields_amount = previous_fields.get(side)
+                if previous_fields_amount is not None:
+                    previous_amount = _derive_currency_amount(previous_fields_amount)
+                    value = _calculate_delta(final_amount, previous_amount)
+                    changed_amount = final_amount
+                    changed_amount["value"] = value
+                    return changed_amount
+                return None
+            changed_amount = final_amount
+            changed_amount["value"] = str(0 - Decimal(changed_amount["value"]))
+            return changed_amount
+    return None
+
+
+def _get_quality(
+    taker_gets: CurrencyAmount,
+    taker_pays: CurrencyAmount,
+) -> str:
+    taker_gets_value = Decimal(taker_gets["value"])
+    taker_pays_value = Decimal(taker_pays["value"])
+    quality = taker_pays_value / taker_gets_value
+    normalized_quality = str(quality.normalize())
+    integer_digit, decimal_places = normalized_quality.split(".")
+    if len(decimal_places) > 15:
+        decimal_places = decimal_places[:15]
```

Comment:
Why is this 15 digits of precision? (I couldn't find a reference which explained the precision to use in this situation)

Reply:
I misinterpreted a comment on Discord. The amount of decimal places are determined based on the [TickSize](https://xrpl.org/ticksize.html) configured by the issuer. If trading an IOU against an IOU the lower TickSize applies.
Considering that it does not make sense to truncate the quality.